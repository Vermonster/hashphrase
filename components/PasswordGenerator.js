import React from 'react';
import { Keyboard, Clipboard, View } from 'react-native';
import loplop from 'loplop';
import FormInput from './FormInput';
import BaseButton from './BaseButton';

export default class PasswordGenerator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      label: '',
      password: '',
    };
  }

    handleChange = name => (e) => {
      const inputValue = e.nativeEvent.text;
      this.setState({ [name]: inputValue });
    };

    addToClipboard = (password) => {
      const { onClipboardSave } = this.props;
      Clipboard.setString(password);
      onClipboardSave();
    };

    generatePassword = (e) => {
      e.preventDefault();
      const { label, password } = this.state;
      this.setState({ generatedPassword: loplop(label, password) }, () => {
        const { generatedPassword } = this.state;
        this.addToClipboard(generatedPassword);
      });
      Keyboard.dismiss();
    };

    render() {
      const { label, password } = this.state;
      return (
        <View>
          <FormInput name="label" value={label} placeholder="Label..." onChanged={this.handleChange('label')} />
          <FormInput name="password" value={password} placeholder="Master Password..." secureTextEntry onChanged={this.handleChange('password')} />
          <BaseButton onPress={this.generatePassword} buttonText="Generate Password" />
        </View>
      );
    }
}
