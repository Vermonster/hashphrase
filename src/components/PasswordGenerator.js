import React from 'react';
import {
  Keyboard, Clipboard, View, StyleSheet, Text,
} from 'react-native';
import { withNamespaces } from 'react-i18next';
import { Button, IconButton } from 'react-native-paper';
import loplop from 'loplop';
import ConfirmationDialog from './ConfirmationDialog';
import FormInput from './FormInput';
import { colors, rowCenter } from '../styles/base';

const styles = StyleSheet.create({
  generatorContainer: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  warning: {
    flexDirection: 'row',
    width: '80%',
    justifyContent: 'space-between',
    marginTop: '-10%',
  },
  icon: {
    marginHorizontal: 0,
  },
});

class PasswordGenerator extends React.Component {
  state = {
    label: '',
    password: '',
    confirmPassword: '',
    generatedPassword: '',
    disabled: true,
    modalVisibility: false,
    inputError: false,
  };

  componentDidUpdate(prevProps) {
    const { isNewPassword } = this.props;
    if (isNewPassword !== prevProps.isNewPassword) {
      this.handleSubmitButtonState();
    }
  }

  showModal = () => this.setState({ modalVisibility: true });

  closeModal = () => this.setState({ modalVisibility: false });

  handleSubmit = (e) => {
    e.preventDefault();
    const { label, password, confirmPassword } = this.state;
    if ((password && confirmPassword) && (password !== confirmPassword)) {
      this.setState({ inputError: true });
    } else {
      this.setState({ inputError: false }, this.generatePassword(label, password));
    }
  };

  addToClipboard = (password) => {
    Clipboard.setString(password);
  };

  handleChange = name => (e) => {
    const inputValue = e.nativeEvent.text;
    this.setState({ [name]: inputValue, inputError: false }, this.handleSubmitButtonState);
  }

  handleSubmitButtonState = () => {
    this.setState({ disabled: true });
    const { label, password, confirmPassword } = this.state;
    const { isNewPassword } = this.props;
    return ((isNewPassword && label && password && confirmPassword)
    || (!isNewPassword && label && password)) ? this.setState({ disabled: false }) : null;
  }

  generatePassword(type, password) {
    this.setState({ generatedPassword: loplop(type, password) }, () => {
      const { generatedPassword } = this.state;
      this.addToClipboard(generatedPassword);
    });
    Keyboard.dismiss();
    this.showModal();
  }

  confirmEntry(inputType) {
    const { label, password, confirmPassword } = this.state;
    let status = null;
    if (inputType === 'label') {
      status = (label.length > 0);
    } else if (inputType === 'password') {
      status = (password.length > 0);
    } else if (inputType === 'confirmPassword') {
      status = (password === confirmPassword);
    }
    this.setState((prevProps) => {
      
    });
    return status;
  }

  render() {
    const {
      generatedPassword, disabled, modalVisibility, inputError, label, password, confirmPassword,
    } = this.state;
    const { isNewPassword, showSnackbar, t } = this.props;

    return (
      <View style={styles.generatorContainer}>
        <FormInput
          value={label}
          componentType="label"
          placeholderText={t('label')}
          prompt={t('passwordLabel')}
          label={t('label')}
          handleChange={this.handleChange}
          onBlur={this.confirmEntry}
        />
        <FormInput
          value={password}
          componentType="password"
          placeholderText={t('password')}
          prompt={t('masterPassword')}
          label={t('password')}
          handleChange={this.handleChange}
        />
        { isNewPassword && (
          <>
            <FormInput
              value={confirmPassword}
              componentType="confirmPassword"
              placeholderText={t('confirmPassword')}
              prompt={t('retypePassword')}
              label={t('confirmPassword')}
              error={inputError}
              handleChange={this.handleChange}
              style={styles.textInput}
            />
            { inputError && (
              <View style={[rowCenter, styles.warning]}>
                <Text style={{ color: colors.warning }}>{t('warning')}</Text>
                <IconButton icon="warning" color={colors.warning} style={styles.icon} />
              </View>
            )
            }
          </>
        )
      }
        <Button
          testID="submit-create-password"
          onPress={this.handleSubmit}
          accessibilityLabel="CREATE ACCOUNT PASSWORD BUTTON"
          disabled={disabled}
          mode="contained"
          dark
        >
          {t('submitMessage')}
        </Button>
        <ConfirmationDialog
          visible={modalVisibility}
          generatedPassword={generatedPassword}
          closeModal={this.closeModal}
          showSnackbar={showSnackbar}
        />
      </View>
    );
  }
}

export default withNamespaces('passwordGenerator')(PasswordGenerator);

export { PasswordGenerator as TestPasswordGenerator };
