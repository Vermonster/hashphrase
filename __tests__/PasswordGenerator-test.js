import React from 'react';
import { shallow } from 'enzyme';
import PasswordGenerator from '../src/components/PasswordGenerator';

describe('<PasswordGenerator />', () => {
  it('should match the existing snapshot', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<PasswordGenerator onClipboardSave={mockFn} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render all of the components', () => {
    const wrapper = shallow(<PasswordGenerator />);
    expect(wrapper.find('FormInput')).toHaveLength(2);
    expect(wrapper.find('BaseButton')).toHaveLength(1);
  });

  it('should set label and password state on handleChange', () => {
    const mockFn = jest.fn();
    const e = { nativeEvent: { text: 'testText!' } };
    const wrapper = shallow(<PasswordGenerator onClipboardSave={mockFn} />);
    const stateValues = ['label', 'password'];

    stateValues.forEach((val) => {
      expect(wrapper.state(val)).toHaveLength(0);
      wrapper.instance().handleChange(val)(e);
      expect(wrapper.state(val)).toBe(e.nativeEvent.text);
    });
  });

  it('should fire the generatePassword function on press', () => {
    const mockEvent = jest.fn();
    const e = { preventDefault: mockEvent };
    const wrapper = shallow(<PasswordGenerator onClipboardSave={mockEvent} />);
    const spy = jest.spyOn(wrapper.instance(), 'generatePassword');
    wrapper.instance().forceUpdate();
    wrapper.find('BaseButton').simulate('press', e);
    expect(spy).toBeCalled();
  });

  it('should fire the addToClipboard function on press', () => {
    const mockEvent = jest.fn();
    const e = { preventDefault: mockEvent };
    const wrapper = shallow(<PasswordGenerator onClipboardSave={mockEvent} />);
    const spy = jest.spyOn(wrapper.instance(), 'addToClipboard');
    wrapper.instance().forceUpdate();
    wrapper.find('BaseButton').simulate('press', e);
    expect(spy).toBeCalled();
  });
});
