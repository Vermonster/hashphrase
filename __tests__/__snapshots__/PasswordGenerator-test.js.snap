// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordGenerator /> should match the existing snapshot 1`] = `
<ForwardRef(View)
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <FormInput
    componentType="label"
    handleChange={[Function]}
    handleInputFocus={[Function]}
    handleInputVisibility={[Function]}
    label="nickname"
    placeholderText={null}
    prompt="nicknameLabel"
    value=""
    visibility={false}
  />
  <FormInput
    componentType="password"
    handleChange={[Function]}
    handleInputFocus={[Function]}
    handleInputVisibility={[Function]}
    inputRef={[Function]}
    label="password"
    placeholderText={null}
    prompt="masterPassword"
    value=""
    visibility={false}
  />
  <withTheme(Button)
    accessibilityLabel="CREATE ACCOUNT PASSWORD BUTTON"
    dark={true}
    disabled={true}
    mode="contained"
    onPress={[Function]}
    style={
      Object {
        "marginTop": 27,
      }
    }
    testID="submit-create-password"
  >
    submitMessage
  </withTheme(Button)>
  <LoadNamespace(ConfirmationDialog)
    clearForm={[Function]}
    closeModal={[Function]}
    generatedPassword=""
    visible={false}
  />
</ForwardRef(View)>
`;
